src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1058 @@
+            edges.push_back(slot_to_sink);
+            cur_costs.push_back(0), cur_caps.push_back(1);
+
+            slot_to_tray.push_back(std::make_pair(i, j));
+        }
+    }
+
+    // run min-cost flow
+    lemon::ListDigraph::ArcMap<int> costs(graph), caps(graph), flow(graph);
+    lemon::ListDigraph::NodeMap<int> labels(graph);
+    lemon::NetworkSimplex<lemon::ListDigraph, int, int> new_graph(graph);
+
+    for (int i = 0; i < static_cast<int>(edges.size()); i++) {
+        costs[edges[i]] = cur_costs[i], caps[edges[i]] = cur_caps[i];
+    }
+
+    labels[src] = -1;
+    for (int i = 0; i < static_cast<int>(nodes.size()); i++) {
+        labels[nodes[i]] = i;
+    }
+    labels[sink] = static_cast<int>(nodes.size());
+
+    new_graph.costMap(costs);
+    new_graph.upperMap(caps);
+    new_graph.stSupply(src, sink, num_flops);
+    new_graph.run();
+    new_graph.flowMap(flow);
+
+    // get, and save, the clustering solution
+    std::vector<std::pair<int, int> > clusters(num_flops);
+    for (lemon::ListDigraph::ArcIt itr(graph); itr != lemon::INVALID; ++itr) {
+        int u = labels[graph.source(itr)], v = labels[graph.target(itr)];
+        if (flow[itr] != 0 && u < num_flops && v >= num_flops) {
+            v -= num_flops;
+            int tray_idx = slot_to_tray[v].first,
+                slot_idx = slot_to_tray[v].second;
+            clusters[u] = std::make_pair(tray_idx, slot_idx),
+            trays[tray_idx].cand[slot_idx] = u;
+        }
+    }
+
+    return clusters;
+}
+
+float MBFF::GetSilh(std::vector<Flop> flops, std::vector<Tray> trays,
+                    std::vector<std::pair<int, int> > clusters) {
+    int num_flops = static_cast<int>(flops.size()), num_trays = static_cast<int>(trays.size());
+
+    float tot = 0;
+    for (int i = 0; i < num_flops; i++) {
+        float min_num = 2000000000;
+        float max_den = MBFF::GetDist(
+            flops[i].pt, trays[clusters[i].first].slots[clusters[i].second]);
+        for (int j = 0; j < num_trays; j++)
+            if (j != clusters[i].first) {
+                max_den = std::max(
+                    max_den, MBFF::GetDist(flops[i].pt,
+                                           trays[j].slots[clusters[i].second]));
+                for (int k = 0; k < static_cast<int>(trays[j].slots.size()); k++) {
+                    min_num = std::min(
+                        min_num, MBFF::GetDist(flops[i].pt, trays[j].slots[k]));
+                }
+            }
+
+        tot += (min_num / max_den);
+    }
+
+    return tot;
+}
+
+// standard K-means++ implementation
+std::vector<std::vector<Flop> > MBFF::KMeans(std::vector<Flop> flops, int K) {
+    int num_flops = static_cast<int>(flops.size());
+
+    std::set<int> chosen;
+    // choose initial center
+    int seed = rand() % num_flops;
+    chosen.insert(seed);
+
+    std::vector<Flop> centers;
+    centers.push_back(flops[seed]);
+
+    float d[num_flops];
```

Comment:
I prefer using C++ containers such as `std::vector` or `std::array` instead of a "raw" array.

Reply:
+1